### PR TITLE
Metric SDK label Enricher interface

### PR DESCRIPTION
**Pull request by jmacd**
This adds an `Enricher` API that supports attaching baggage attributes as metric labels.
This was a core feature of OpenCensus that OTel metrics hasn't implemented here, and we think this support will be specified for the default SDK.
Inspired by #1271 (@hstan) and my comments there.

This PR is a draft until #1378 merges, after which I'll add a setting to the controller and an example.

### CODE REVIEW

**Comment by jmacd** [comment]
@hstan I hope you don't mind that I experimented around the topic of your PR #1271. I ended up writing a few more comments on the behavior of the enricher, see what you think.

@open-telemetry/specs-metrics-approvers This is important for the OpenCensus vision. Please review.

File: sdk/metric/sdk.go [hunk 1]
```diff
@@ -352,6 +359,20 @@ func (m *Accumulator) Collect(ctx context.Context) int {
...
+	if m.enricher == nil {
+		return kvs
+	}
+	out, err := m.enricher(ctx, kvs)
+
+	// Return the enricher result if it is non-nil and no error.
+	if err == nil && out != nil {
+		return out
+	}
+	otel.Handle(err)
```
Comment (punya) [inline sdk/metric/sdk.go:372]: If `out == nil`, we will invoke `otel.Handle(nil)`. I couldn't find any documentation for what `otel.Handle` is supposed to do in that case.

```suggestion
	if err != nil {
		otel.Handle(err)
		return kvs
	}
	
	if out == nil {
		return kvs
	}
	
	return out
```

**Review by punya — COMMENTED** [review]
A few small localized comments, I am still learning about the overall codebase.

**Comment by punya** [comment]
@jmacd my apologies, please feel free to disregard my review comments if you wish - I somehow failed to notice that this is a draft.

**Comment by jmacd** [comment]
@bogdandrutu Please review.

**Comment by jmacd** [comment]
I will re-open this later if needed, closing this as we have #1271.

**Pull request closed without merging**